src/util/small_map.h

```diff
@@ -0,0 +1,472 @@
+/*******************************************************************\
+
+Module: Small map
+
+Author: Daniel Poetzl
+
+\*******************************************************************/
+
+/// \file
+/// Small map
+
+#ifndef CPROVER_UTIL_SMALL_MAP_H
+#define CPROVER_UTIL_SMALL_MAP_H
+
+#include <bitset>
+#include <utility>
+#include <type_traits>
+#include <memory>
+#include <limits>
+#include <cstdint>
+#include <cstring>
+
+#include "invariant.h"
+
+/// Map from small integers to values
+///
+/// A data structure that maps small integers (in {0, ..., Num-1}) to values.
+/// It is designed to be more memory-efficient than std::map for this use case.
```

Comment:
How much "more" memory-efficient is it? And what values of `Num` are typically to be used?

Reply:
We're using the small map to represent the children of a node in a tree data structure. A node can have up to `Num` children, with indices `0...Num-1`, but not all of them are usually present.

Typical values for `Num` are say 8 or 16 (we're using 8 at the moment). For a map that stores pointers (as in our case) it uses about 1/5 of the memory compared to `std::map`.

Reply:
Ok, cool, it would be nice to have as concrete as possible data points in the comments. Statements like "for Num=8, std::map requires X bytes on a 64-bit system when using GNU STL while this map only requires Y bytes".

Reply:
Ok, I've added the data about the memory consumption to the class description.